src/Core/Services/CryptoService.cs

```diff
@@ -389,28 +389,57 @@ public async Task ToggleKeyAsync()
             await SetKeyAsync(key);
         }
 
-        public async Task<SymmetricCryptoKey> MakeKeyAsync(string password, string salt,
-            KdfType? kdf, int? kdfIterations)
+        public async Task<SymmetricCryptoKey> MakeKeyAsync(string password, string salt, KdfConfig kdfConfig)
         {
             byte[] key = null;
-            if (kdf == null || kdf == KdfType.PBKDF2_SHA256)
+            if (kdfConfig.Type == null || kdfConfig.Type == KdfType.PBKDF2_SHA256)
             {
-                if (kdfIterations == null)
+                if (kdfConfig.Iterations == null)
                 {
-                    kdfIterations = 5000;
+                    kdfConfig.Iterations = 5000;
                 }
-                if (kdfIterations < 5000)
+                if (kdfConfig.Iterations < 5000)
                 {
                     throw new Exception("PBKDF2 iteration minimum is 5000.");
                 }
                 key = await _cryptoFunctionService.Pbkdf2Async(password, salt,
-                    CryptoHashAlgorithm.Sha256, kdfIterations.Value);
+                    CryptoHashAlgorithm.Sha256, kdfConfig.Iterations.Value);
             }
-            else if (kdf == KdfType.Argon2id)
+            else if (kdfConfig.Type == KdfType.Argon2id)
             {
-                var iterations = kdfIterations.Value;
-                const int parallelism = 1;
-                const int memory = 1024 * 16; // 16 MiB
+                if (kdfConfig.Iterations == null)
+                {
+                    kdfConfig.Iterations = Constants.Argon2Iterations;
+                }
+                if (kdfConfig.Iterations < 2)
+                {
+                    throw new Exception("Argon2 iterations minimum is 2");
+                }
+
+                if (kdfConfig.Memory == null)
+                {
+                    kdfConfig.Memory = Constants.Argon2Memory;
+                }
+                if (kdfConfig.Memory < 16)
+                {
+                    throw new Exception("Argon2 memory minimum is 16 MB");
+                } else if (kdfConfig.Memory > 1024)
+                {
+                    throw new Exception("Argon2 memory maximum is 1024 MB");
+                }
+
+                if (kdfConfig.Parallelism == null)
+                {
+                    kdfConfig.Parallelism = Constants.Argon2Parallelism;
+                } 
+                else if (kdfConfig.Parallelism < 1)
+                {
+                    throw new Exception("Argon2 parallelism minimum is 1");
+                }
+
+                var iterations = kdfConfig.Iterations.GetValueOrDefault(Constants.Argon2Iterations);
+                var memory = kdfConfig.Memory.GetValueOrDefault(Constants.Argon2Memory) * 1024;
+                var parallelism = kdfConfig.Parallelism.GetValueOrDefault(Constants.Argon2Parallelism);
```

Comment:
There is a repetition to fill the values if the original is `null` (with the if and with `GetValueOrDefault(...)`).
So I'd just fill the local variables first and then check against them or remove the local variables and use the updated `kdfConfig` when calling the service.
Also there is a formatting issue on line 426: `} else if (kdfConfig.Memory > 1024)` the `else if` should be in the next line.

Reply:
Fixed in [5804842](https://github.com/bitwarden/mobile/pull/2328/commits/58048427009097c3497bddd92952c27bfc5490e0)